.github/workflows/docs.yml

```diff
@@ -2,34 +2,52 @@ name: Docs
 
 on:
   push:
-    branches: [main]
+    branches:
+      - main
+      - "*"
 
 concurrency:
-  group: "docs"
+  group: "docs-${{ github.ref_name }}"
   cancel-in-progress: false
 
 jobs:
+  build:
+    runs-on: ubuntu-latest
+    permissions:
+      pages: write
+      id-token: write
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v4
 
-    build:
-      runs-on: ubuntu-latest
-      permissions:
-        pages: write
-        id-token: write
-      steps:
-      - name: Checkout repository
-        uses: actions/checkout@v4
-      - name: Setup Anchor
-        uses: ./.github/actions/anchor
-      - name: Setup Github Pages
-        uses: actions/configure-pages@v5
-      - name: Install dependencies
-        run: yarn install
-      - name: Build docs
-        run: yarn build docs/whirlpool --output-style static
-      - name: Upload artifact
-        uses: actions/upload-pages-artifact@v3
-        with:
-          path: ./docs/whirlpool/dist
-      - name: Deploy artifact
-        id: deployment
-        uses: actions/deploy-pages@v4
+    - name: Setup Anchor
+      uses: ./.github/actions/anchor
+
+    - name: Setup Github Pages
+      uses: actions/configure-pages@v5
+
+    - name: Install dependencies
+      run: yarn install
+
+    - name: Build docs
+      run: yarn build docs/whirlpool --output-style static
+
+    - name: Upload artifact
+      uses: actions/upload-pages-artifact@v3
+      with:
+        path: ./docs/whirlpool/dist
+
+    - name: Deploy artifact
+      if: github.ref_name == 'main'
+      id: deployment
+      uses: actions/deploy-pages@v4
+      with:
+        branch: gh-pages
+
+    - name: Deploy preview
+      if: github.ref_name != 'main'
+      run: |
+        PREVIEW_DIR="previews/${GITHUB_REF_NAME##*/}"
```

Comment:
How exactly does this work? All I see this doing is creating a new folder and moving the docs output folder there